storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2156,11 +2156,99 @@ static bool innodb_init_param(void) {
   return (TRUE);
 }
 
+/** Scan the sdi id from mysql.tables
+@param[in]	name 		Name of table database/name
+@param[out]     sdi_id		id of table
+@param[in]	thd 		THD */
+void get_id_from_mysql_table_scan(const std::string name, uint64 *id,
+                                  THD *thd) {
+  std::string db_name;
+  std::string table_name;
+  uint64 schema_id = 0;
+
+  dict_name::get_table(name, db_name, table_name);
+
+  ut_ad(db_name.compare("mysql") != 0);
+
+  dict_table_t *sys_tables = nullptr;
+  btr_pcur_t pcur;
+  const rec_t *rec = nullptr;
+  mtr_t mtr;
+  MDL_ticket *mdl = nullptr;
+  mem_heap_t *heap = mem_heap_create(1000);
+
+  mutex_enter(&(dict_sys->mutex));
+
+  auto build_mysql_schema = [&]() {
+    std::map<std::string, int> mysql_schema;
+
+    mtr_start(&mtr);
+    rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/schemata",
+                              &sys_tables);
+    while (rec) {
+      uint64 rec_schema_id;
+      std::string rec_name;
+
+      dd_process_schema_rec(heap, rec, sys_tables, &mtr, &rec_name,
+                            &rec_schema_id);
+      mysql_schema.insert(std::pair<std::string, int>(rec_name, rec_schema_id));
+      mem_heap_empty(heap);
+
+      mtr_start(&mtr);
+      rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+    }
+
+    mtr_commit(&mtr);
+    dd_table_close(sys_tables, thd, &mdl, true);
+    mem_heap_empty(heap);
+
+    return mysql_schema;
+  };
+
+  static auto mysql_schema = build_mysql_schema();
+
+  schema_id = mysql_schema[db_name];
+  ut_a(schema_id != 0);
+
+  mtr_start(&mtr);
+
+  rec =
+      dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/tables", &sys_tables);
+
+  while (rec) {
+    uint64 rec_schema_id;
+    std::string rec_name;
+    uint64 rec_id;
+
+    dd_process_dd_tables_rec(heap, rec, sys_tables, &mtr, &rec_schema_id,
+                             &rec_name, &rec_id);
+    mem_heap_empty(heap);
+
+    if (schema_id == rec_schema_id && table_name.compare(rec_name) == 0) {
+      *id = rec_id;
+      break;
+    }
+
+    mtr_start(&mtr);
+    rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+  }
+
+  ut_a(*id != 0);
+
+  dd_table_close(sys_tables, thd, &mdl, true);
+  mem_heap_empty(heap);
```

Comment:
mtr_commit() missing here. Incase we traverse all DD records and reach here (without the break)

Reply:
We will assert in that case. It is not expected to happpen.

Reply:
A debug assert is fine. Release asserts should be used judiciously. Infact, I see that you used lot of ut_a().

Here please try to add mtr_commit and return error

Reply:
Added exit with message in case table not found in mysql/schemata or mysql/tables